xinference/core/worker.py

```diff
@@ -62,29 +75,53 @@ async def __pre_destroy__(self):
     def get_model_count(self) -> int:
         return len(self._model_uid_to_model)
 
-    def _choose_subpool(self) -> str:
-        min_running_model_count = None
-        target_subpool_address = None
-        for subpool_address in self._subpool_address_to_model_uids:
-            running_model_count = len(
-                self._subpool_address_to_model_uids[subpool_address]
-            )
-            if (
-                min_running_model_count is None
-                or running_model_count < min_running_model_count
-            ):
-                min_running_model_count = running_model_count
-                target_subpool_address = subpool_address
-
-        if target_subpool_address:
-            logger.debug(
-                "Subpool selected: %s, running model count: %d",
-                target_subpool_address,
-                min_running_model_count,
+    def allocate_devices(self, n_gpu: int) -> List[int]:
+        """
+        Allocate GPUs to the model based on the form-filling method to achieve a balanced GPU load as much as possible.
+        """
+        allocated_device: Dict[int, int] = {}
+        devices: List[int] = []
+        for _ in range(n_gpu):
+            min_running_model_count = min(
+                [
+                    len(models) + allocated_device.get(dev, 0)
+                    for dev, models in self._gpu_to_model_uids.items()
+                ]
             )
-            return target_subpool_address
-
-        raise RuntimeError("No available slot found")
+            for dev in self._gpu_to_model_uids:
+                running_model_count = len(
+                    self._gpu_to_model_uids[dev]
+                ) + allocated_device.get(dev, 0)
+                if (
+                    len(devices) < n_gpu
+                    and running_model_count <= min_running_model_count
+                ):
+                    min_running_model_count = running_model_count
+                    devices.append(dev)
+                    allocated_device[dev] = allocated_device.get(dev, 0) + 1
+        return sorted(devices)
+
+    async def _create_subpool(
+        self,
+        model_uid: str,
+        n_gpu: Optional[Union[int, str]] = "auto",
+    ) -> Tuple[str, List[str]]:
+        env = {}
+        devices = []
+        if isinstance(n_gpu, int) or (n_gpu == "auto" and cuda_count() > 0):
```

Comment:
This is an user input parameter. I can't assume user input for any situation. The check here is just in case. `n_gpu` should never exceeded GPUs on the machine.

Reply:
> Yeah, but you should let the user know that the available device is not able to satisfy the requirement by raising an Error.

This is done by parameter checking otherwhere. I'm very confused what is the relevance of the code you labeled here to the problem you're talking about. The fact that the code goes here must mean that the number of requested devices is less than or equal to the number of available devices.

Reply:
Let me know the line number of `n_gpu` validation.

Reply:
Oh I see your changes on allocate_devices.